rmw_fastrtps_shared_cpp/include/rmw_fastrtps_shared_cpp/custom_subscriber_info.hpp

```diff
@@ -204,12 +196,26 @@ class SubListener : public EventListenerInterface, public eprosima::fastdds::dds
   bool incompatible_qos_changes_
   RCPPUTILS_TSA_GUARDED_BY(on_new_event_m_);
 
+  eprosima::fastdds::dds::SubscriptionMatchedStatus matched_status_
+  RCPPUTILS_TSA_GUARDED_BY(on_new_event_m_);
+
+  bool matched_changes_
+  RCPPUTILS_TSA_GUARDED_BY(on_new_event_m_);
+
+  eprosima::fastdds::dds::SubscriptionMatchedStatus unmatched_status_
```

Comment:
we don't need `matched_status_` and `unmatched_status_`, we should only have one

Reply:
We have 2 events.  
While user takes status for matched event, we can clear matched status (But we should not clear unmatched status). 

```c++
typedef struct RMW_PUBLIC_TYPE rmw_matched_status_s
{
  ...

  /**
   * The change in current_count since the last time the status was read.
   */
  int32_t current_count_change;
} rmw_matched_status_t;
```

current_count _change means `The change in current_count since the last time the status was read`.  So I think the change should be from the last matched event status read. This read should not affect unmatched status.  

So we use 2 status separately for matched event and unmatched event.

Reply:
i.e. an event that's triggered when the "current_count_change" increased or decreased (or even when it's zero).

Reply:
If we only use matched event, user cannot be notified while unmatched event.  

User can set callback via rmw_event_set_callback().  
The callback is defined as the below 
```c
/// Common event callback type signature.
/**
 * Event callbacks of this type can be called in various scenarios, e.g.
 * data becomes available on a subscription, a QoS event has occurred, or
 * something similar.
 *
 * The user_data argument is given by the user when registering the callback,
 * and is given back to the callback each time so it can have associated,
 * user-defined state.
 *
 * The number_of_events argument indicates the number of events since the
 * callback was called.
 * This is most often 1, but can be > 1 when events occur before the callback
 * is registered.
 * It should never be 0.
 *
 * \sa rmw_subscription_set_on_new_message_callback()
 * \sa rmw_service_set_on_new_request_callback()
 * \sa rmw_client_set_on_new_response_callback()
 * \sa rmw_event_set_callback()
 */
typedef void (* rmw_event_callback_t)(const void * user_data, size_t number_of_events);
```
number_of_events should be bigger than 0.  We cannot use negative number for unmatched. Do you want to change this interface to allow negative number ?

Reply:
That seems to be a problem in the events executor, I'm not sure why the data that it's passed to the callback in the case of an event is only a number, instead of the full event status struct.
I wouldn't try to modify how the event should be defined to try to fit that.

Reply:
I agree with @ivanpauno. May we always pass `1` as the `number_of_events`?

It will mean something like. The matching status has changed `1` times.

Reply:
According to current implementation, there are 2 ways

1. Register callback to DDS
`rmw_event_callback_t` is only used for rmw_event_callback_t() which register callback directly to DDS.   
That is, while the event occurs, the callback function is called at DDS layer.   

The event number always set to 1 while calling callback.
e.g.  
https://github.com/ros2/rmw_fastrtps/blob/ccea7a356900a16a2872670b67952c99a72ab65d/rmw_fastrtps_shared_cpp/src/custom_publisher_info.cpp#L213-L220

2. Wait event and take status
Full event status can be gotten via rmw_take_event(). It is used to wait event via wait_set(). And Once event occur, get event status via rmw_take_event() and call user's callback function with taken data.

Reply:
@ivanpauno Friendly ping.  
Whether do we consider only one matched event ?   
If we only use one matched event, while `rmw_event_callback_t` is called, user cannot directly distinguish matched/unmatched event. User have to use rmw_take_event() to further check.

Reply:
> If we only use one matched event, while rmw_event_callback_t is called, user cannot directly distinguish matched/unmatched event. User have to use rmw_take_event() to further check.

Yes, that will be a problem with the event callback API.

Reply:
> Yes, that will be a problem with the event callback API.

So we keep current design which use matched/unmatched 2 events.    
Besides, I consider changing rmw_matched_status_t to make more clear definition for user.  
https://github.com/ros2/rmw/pull/331#issuecomment-1321135801  
What do you think ?

Reply:
> So we keep current design which use matched/unmatched 2 events.

As commented above, I would prefer to not have both.
It seems that we're forcing them to be different events, but in reality they're the same thing.
As @MiguelCompany commented, you can use 1 in number of events.
The user would have to get the event status to know the actual info when using callbacks.

Reply:
@wjwwood do you have any opinion on this?

Reply:
> As commented above, I would prefer to not have both.
> It seems that we're forcing them to be different events, but in reality they're the same thing.
> As @MiguelCompany commented, you can use 1 in number of events.
> The user would have to get the event status to know the actual info when using callbacks.

I got your thourght.   
If the user wants to get unmatched event, the case of only matched event will make the user inconvenient to use. He have to do further check if current event is match or unmatch.  Users who do not know the DDS specification will wonder why it was designed this way.